host/sgx/exception_sim.c

```diff
@@ -0,0 +1,226 @@
+    uint32_t sgx_vector = OE_EXCEPTION_UNKNOWN;
+    switch (sig_num)
+    {
+        case SIGFPE: // OE_EXCEPTION_DIVIDE_BY_ZERO
+            sgx_vector = 0;
+            break;
+        case SIGTRAP: // OE_EXCEPTION_BREAKPOIN
+            sgx_vector = 3;
+            break;
+        case SIGILL: // OE_EXCEPTION_ILLEGAL_INSTRUCTION
+            sgx_vector = 6;
+            break;
+        case SIGSEGV: // OE_EXCEPTION_ACCESS_VIOLATION
+            sgx_vector = 13;
+            break;
+        case SIGBUS: // OE_EXCEPTION_MISALIGNMENT
+            sgx_vector = 17;
+            break;
+    }
+
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+    // Check struct sgx_exit_info for the detail that lhs is 8 bits only.
+    ssa_gpr->exit_info.as_fields.vector = (uint8_t)sgx_vector;
+}
+
+static void _update_context_from_ssa(ucontext_t* context, sgx_tcs_t* tcs)
+{
+    printf("__LINE__: %d\n",__LINE__);//
+    sgx_ssa_gpr_t* ssa_gpr = _get_ssa_gpr(tcs);
+
+    context->uc_mcontext.gregs[REG_RAX] = (greg_t)ssa_gpr->rax;
+    context->uc_mcontext.gregs[REG_RBX] = (greg_t)ssa_gpr->rbx;
+    context->uc_mcontext.gregs[REG_RCX] = (greg_t)ssa_gpr->rcx;
+    context->uc_mcontext.gregs[REG_RDX] = (greg_t)ssa_gpr->rdx;
+    context->uc_mcontext.gregs[REG_RSP] = (greg_t)ssa_gpr->rsp;
+    context->uc_mcontext.gregs[REG_RBP] = (greg_t)ssa_gpr->rbp;
+    context->uc_mcontext.gregs[REG_RSI] = (greg_t)ssa_gpr->rsi;
+    context->uc_mcontext.gregs[REG_RDI] = (greg_t)ssa_gpr->rdi;
+    context->uc_mcontext.gregs[REG_R8] = (greg_t)ssa_gpr->r8;
+    context->uc_mcontext.gregs[REG_R9] = (greg_t)ssa_gpr->r9;
+    context->uc_mcontext.gregs[REG_R10] = (greg_t)ssa_gpr->r10;
+    context->uc_mcontext.gregs[REG_R11] = (greg_t)ssa_gpr->r11;
+    context->uc_mcontext.gregs[REG_R12] = (greg_t)ssa_gpr->r12;
+    context->uc_mcontext.gregs[REG_R13] = (greg_t)ssa_gpr->r13;
+    context->uc_mcontext.gregs[REG_R14] = (greg_t)ssa_gpr->r14;
+    context->uc_mcontext.gregs[REG_R15] = (greg_t)ssa_gpr->r15;
+    context->uc_mcontext.gregs[REG_RIP] = (greg_t)ssa_gpr->rip;
+    context->uc_mcontext.gregs[REG_EFL] = (greg_t)ssa_gpr->rflags;
+}
+
+uint64_t oe_host_handle_exception_sim(ucontext_t* context, int sig_num)
+{
+    void* enclave_fsbase = oe_get_fs_register_base();
+    oe_sgx_td_t* td = (oe_sgx_td_t*)enclave_fsbase;
+    sgx_tcs_t* tcs =
+        (sgx_tcs_t*)((unsigned char*)td - OE_TD_FROM_TCS_BYTE_OFFSET);
+    uint64_t ret = OE_EXCEPTION_CONTINUE_SEARCH;
+
+    uint64_t enclave_start, enclave_end, rip;
+    oe_enclave_t* enclave = oe_query_enclave_instance((void*)tcs);
+    if (enclave == NULL)
+    {
+        goto done;
+    }
+
+    enclave_start = enclave->addr;
+    enclave_end = enclave->addr + enclave->size;
+    rip = (uint64_t)(context->uc_mcontext.gregs[REG_RIP]);
+    if (rip >= enclave_start && rip < enclave_end)
+    {
+        // Simulate the AEX in SGX hardware mode.
+        // Copy the data of context into ssa manually.
+        _oe_aex_sim(tcs);
```

Comment:
There's actually a lot of stuff that AEX does which probably is not directly relevant to exception handling, but it makes me nervous because I don't understand the degree of accuracy it needs to be simulated at to not cause other kinds of issues downstream. A couple of things off the top of my head:

* AEX preserves XSAVE state in addition to GPRs. 
* AEX sets the context to a synthetic state that the host sees.

@mingweishih if he knows of any specific implications.